### PR TITLE
add s3 backup and restore to the rke-tools 

### PR DESCRIPTION
**Pull request by guangbochen**
Take etcd sanpshot and saved to s3:
```
etcd-backup save --once --debug --s3-backup=true --s3-endpoint=s3.amazonaws.com --s3-accessKey=xxx --s3-secretKey=xxx --s3-bucketName=my-bucket-name
```

Download etcd snapshot from s3:
```
etcd-backup download --s3-accessKey=xxx --s3-secretKey=xxx --s3-bucketName=my-bucketname --s3-endpoint=s3.amazonaws.com --endpoints=192.168.99.100:2379 --name="rke_etcd_snapshot_2018-11-19T11:03:05+08:00"
```

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -37,53 +121,41 @@ func main() {
...
-			cli.BoolFlag{
-				Name:   "debug",
-				Usage:  "Verbose logging information for debugging purposes",
-				EnvVar: "RANCHER_DEBUG",
-			},
-			cli.BoolFlag{
-				Name:  "once",
-				Usage: "Take backup only once",
+		Name:  "rke-tools",
+		Usage: "Perform rke tools",
```
Comment (galal-hussein) [inline main.go]: I am not sure i understand the name here, can we at least add more descriptive usage message for the whole tool, this should be a tool to perform backup operations on etcd

File: main.go [hunk 2]
```diff
@@ -37,53 +121,41 @@ func main() {
...
-				Name:   "cert",
-				Usage:  "Etcd client certificate path",
-				EnvVar: "ETCD_CERT",
-			},
-			cli.StringFlag{
-				Name:   "key",
-				Usage:  "Etcd client key path",
-				EnvVar: "ETCD_KEY",
+			{
+				Name:   "restore",
```
Comment (galal-hussein) [inline main.go]: this subcommand is not exactly restore, as far as i understand we are saving/downloading the backup from s3

**Comment by moelsayed** [comment]
@guangbochen Can you please push the vendor code in a separate commit ?

**Comment by guangbochen** [comment]
@alena1108 resolved for the above reviews, thanks.

File: main.go [hunk 3]
```diff
@@ -107,32 +169,63 @@ func RollingBackupAction(c *cli.Context) error {
...
 		"creation":  creationPeriod,
 		"retention": retentionPeriod,
 	}).Info("Initializing Rolling Backups")
 
+	s3Backup := c.Bool("s3-backup")
+	s3Config := &s3Server{
+		Backup:     s3Backup,
+		Endpoint:   c.String("s3-endpoint"),
+		AccessKey:  c.String("s3-accessKey"),
+		SecretKey:  c.String("s3-secretKey"),
```
Comment (moelsayed) [inline main.go]: We shouldn't log sensitive information like secret or private keys

File: main.go [hunk 4]
```diff
@@ -239,3 +339,94 @@ func DeleteBackup(file os.FileInfo) {
...
+	return svc, nil
+}
+
+func uploadBackupFile(svc *minio.Client, bucketName, fileName, filePath string) error {
+	// Upload the zip file with FPutObject
+	log.Infof("invoking uploading backup file %s to s3", fileName)
+	for retries := 0; retries <= s3ServerRetries; retries++ {
+		n, err := svc.FPutObject(bucketName, fileName, filePath, minio.PutObjectOptions{ContentType: contentType})
+		if err != nil {
+			fmt.Printf("failed to upload etcd snapshot file: %v, retried %d times", err, retries)
```
Comment (moelsayed) [inline main.go]: let's make this a log statement, info or warn

File: main.go [hunk 5]
```diff
@@ -239,3 +339,94 @@ func DeleteBackup(file os.FileInfo) {
...
+	fileName := c.String("name")
+	if len(fileName) == 0 {
+		return fmt.Errorf("empty file name")
+	}
+
+	log.Infof("invoking downloading backup files: %s", fileName)
+	for retries := 0; retries <= s3ServerRetries; retries++ {
+		err := client.FGetObject(s3Config.BucketName, fileName, backupBaseDir+"/"+fileName, minio.GetObjectOptions{})
+		if err != nil {
+			fmt.Printf("failed to download etcd snapshot file: %v, retried %d times", err, retries)
```
Comment (moelsayed) [inline main.go]: Should also be a log.

**Pull request merged**